specs/tables.md

```diff
@@ -356,3 +356,72 @@ The table below lists all of copy pairs supported in the copy table:
+3. `base_limb`: 64-bit limb representing the integer base of the exponentiation operation.
+4. `exponent_lo_hi`: 128-bit low/high parts of an intermediate value that starts at the integer exponent.
+5. `exponentiation_lo_hi`: 128-bit low/high parts of an intermediate value that starts at the result of the exponentation operation.
+
+The lookup entry is not a single row in the table, and not every row corresponds to a lookup entry. Instead, a lookup entry is constructed from the first 4 rows in each exponentiation event. For simplicity in the `specs` implementation, we combine all those rows into a single row. But in the `circuits` implementation, we try to lower the number of columns in exchange of increased number of rows.
+
+Depending on the value of the `exponent` within the exponentiation operation, the `EXP` gadget will either:
+1. Do no lookup if `exponent == 0` since `base ^ 0 == 1 (mod 2^256)`
+2. Do no lookup if `exponent == 1` since `base ^ 1 == base (mod 2^256)`
+3. Do 1 lookup to a row if `exponent == 2` since there is a single step in the exponentiation trace, i.e. `base ^ 2 == base * base (mod 2^256)`, implying that `is_first == is_last == 1` for this row.
+4. Do 2 lookups to 2 different rows if `exponent > 2` since there are more than one steps in the exponentiation trace, i.e. a lookup to `is_last == 0` and a lookup to `is_last == 1`.
+
+Consider `3 ^ 13 == 1594323 (mod 2^256)`. The exponentiation table assignment looks as follows:
+
+| identifier | is_last | base_limb0 | base_limb1 | base_limb2 | base_limb3 | exponent_lo | exponent_hi | lsb_exponent | exponentiation_lo | exponentiation_hi |
+|------------|---------|------------|------------|------------|------------|-------------|-------------|--------------|-------------------|-------------------|
+| $rwc       | 0       | 3          | 0          | 0          | 0          | 13          | 0           | 13           | 1594323           | 0                 |
+| $rwc       | 0       | 3          | 0          | 0          | 0          | 12          | 0           | 12           | 531441            | 0                 |
+| $rwc       | 0       | 3          | 0          | 0          | 0          | 6           | 0           | 6            | 729               | 0                 |
+| $rwc       | 0       | 3          | 0          | 0          | 0          | 3           | 0           | 3            | 27                | 0                 |
+| $rwc       | 1       | 3          | 0          | 0          | 0          | 2           | 0           | 2            | 9                 | 0                 |
+
+For `exponent == 13`, i.e. scenario #4 we do two lookups:
+1. Lookup to first row:
+```
+Row(identifier=rwc, is_last=0, base_limbs=[3, 0, 0, 0], exponent_lo_hi=[13, 0], lsb_exponent=13, exponentiation_lo_hi=[1594323, 0])
+```
+2. Lookup to last row:
+```
+Row(identifier=rwc, is_last=1, base_limbs=[3, 0, 0, 0], exponent_lo_hi=[2, 0], lsb_exponent=2, exponentiation_lo_hi=[9, 0])
```

Comment:
Why do we need this lookup? It seems the first lookup is already enough for the EXP opcode. What would happen if we don't do this lookup?

Reply:
We wish to validate that the exponentiation trace finally reduces to `exponent == 2`, i.e. `a * a == a^2 (mod 2^256)`.

In the exponentiation circuit, we can only verify that:
```python
if is_last is 1 then assert exponent == 2
```

But we cannot ensure that such a row occurs. In order to ensure that this row occurs, we do a lookup from the `EXP` gadget.

Reply:
makes sense. Thanks for the explanation.